src/historywork/RunCommandWork.h

```diff
@@ -8,23 +8,28 @@
 
 namespace stellar
 {
+using RunCommandInfo = std::pair<std::string, std::string>;
 
-// This subclass exists for two reasons: first, to factor out a little code
-// around running commands, and second to ensure that command-running
-// happens from onStart rather than onRun, and that onRun is an empty
-// method; this way we only run a command _once_ (when it's first
-// scheduled) rather than repeatedly (racing with other copies of itself)
-// when rescheduled.
+/**
+ * This class helps run various commands, that require
+ * process spawning. This work is not scheduled while it's
+ * waiting for a process to exit, and wakes up when it's ready
+ * to be scheduled again.
+ */
 class RunCommandWork : public Work
```

Comment:
Regarding comments on `Work` vs `BasicWork`: I like the idea of keeping the work tree consistent and implementers only deriving from Work.

Reply:
Thinking about this more, I think this is related to a `FIXME` in `VerifyBucketWork::doWork`. Let's say we follow your suggestion - this means that outside work won't be allowed to `addWork` to some other works (that inherit directly from `BasicWork`, for example). In the `VerifyBucketWork` example, `DownloadBucketsWork` won't be allowed to add chains of GetRemoteFileWork->VerifyBucketWork, because `GetRemoteFileWork` would inherit from `RunCommandWork` (and hence from `BasicWork`). I think this pushes for a different mindset regarding how we add work. We either make parent `DownloadBucketsWork` worry about when to properly dispatch download and verify works (creating a flatter structure, as both download and verify works would be children of DownloadBucketsWork), or create an abstraction for "download and verify a bucket", and it would be a child of `DownloadBucketsWork`.

Reply:
I have also given this topic more thought. Both of my earlier comments here were viewing this as an implementation issue and not a design issue. After considering your latest comment, it is clear to me that this is in fact a design issue. The root cause of this issue is that `addWork` is part of the public interface of `Work`. I do not believe that this is desirable, since it blurs the meaning of child work.

Suppose that C is a child work of W. Does this imply that W depends on the output of C, or that W is composed from C? The first would indicate that the relationship between W and C is purely one of ordering: C must complete before W can proceed. The second would indicate that W manages C. For me, the answer is the second. The FIXME in `VerifyBucketWork::doWork` is a reflection of the logical issues associated with using the child work relationship to model a dependence on output: the derived class is not aware that it has any children (since the call to `addWork` is not virtual and has no mechanism to inform the derived class of the new child), yet it still must wait for them to finish. Specifically, when `VerifyBucketWork` is used as part of `DownloadBucketsWork`, we observe that it is-implemented-in-terms-of `GetAndUnzipRemoteFileWork` yet it is unaware of this relationship. (For more information on is-implemented-in-terms-of, see items 38 and 39 of Scott Meyers “Effective C++” Third Edition.)

The relationship is-implemented-in-terms-of should in this case be modeled using explicit composition. `VerifyBucketWork` is already an implicit composition, since the base class `Work` is managing the relationship to `GetAndUnzipRemoteFileWork`. To fix this, `addWork` should be `protected` rather than `public`. Then the derived class will know about all of its children, since it will have added them. In this case, the `VerifyBucketWork` concept could be broken down into its components:
- Get (and unzip) the remote file
- Verify the bucket

The first step is already implemented as `GetAndUnzipRemoteFileWork` (itself a composite work, as the name indicates). The second part could be implemented as `VerifyBucketWork` which would now only verify a bucket that already exists on disk. The composite could be named `DownloadAndVerifyBucketWork` which would first execute `GetAndUnzipRemoteFileWork`, then execute `VerifyBucketWork`. In this approach, there is no need for any external calls to `addWork`. An added benefit of this approach is that it will encourage implementors to distill work down to largely independent components, which should lead to much simpler and more modular implementations. It should be noted that this will come at the cost of implementors of composite work classes needing to manually manage the execution order among their children (some composite work classes are already implemented in this manner).

Of course, there exist types of work where new work should be added externally. For example, it should always be possible to add arbitrary child work to the `WorkScheduler`, since it is a top-level object only responsible for the scheduling of whatever work is added as its child. This could be achieved by adding `WorkScheduler::addWork` to the public interface of `WorkScheduler` which simply forwards to the protected function `Work::addWork`. More sophisticated examples could be built in an analogous manner, by exposing functions in the public interface which only allow a limited forwarding to `Work::addWork` (for example, the work to add must be of a certain type chosen by the class but the caller can choose the parameters).

Reply:
@MonsieurNicolas any thoughts about the proposal in the previous comment?

Reply:
To add to what Jon posted and my previous statement that moving to this design comes with a cost for implementers, I'd like to add a few more thoughts:
* This means that Work would need to check on status of each child, and react appropriately to each possible state of the child (before if parent got to `onRun`, all of its children were guaranteed to be in successful state). 
* A positive thing I found about this is that parent can control if it wants to keep going (currently if a single child fails, parent fails by default). This would resolve issues like this one https://github.com/stellar/stellar-core/issues/1649, for example.

Reply:
yeah I agree: using parent/child this way is not desirable. The only reason for deriving from `RunCommand` for example, would be to make some new kind of `RunCommand`, but in our code base I don't think we ever need this (as all we do is sequencing, maybe `batchWork` falls under a legitimate case of derivation).
So in general, we should derive only from `Work` or `BasicWork`.

As for `addWork`: yeah we can make it protected to force proper composition.
In some cases though, instead of creating single use `BasicWork` (or `Work`) objects, it might be enough to just perform the sequencing directly in `doWork` based on booleans/state tracking variable.

As for the other questions...

> This means that Work would need to check on status of each child, and react appropriately to each possible state of the child (before if parent got to onRun, all of its children were guaranteed to be in successful state).

yes, this is exactly what we discussed before. The sequencing is explicit in every `doWork` implementation, which is why providing some helper functions like `allChildrenSuccessful()` can help avoid writing the same checks everywhere. The more complicated `Work` already follows this pattern anyways.